### PR TITLE
Integrate MSE notifications into PacketFence

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## Description

Integrate MSE notifications into PacketFence
## Dependencies

Depends on #1577 for the extendable profile filters
## NEWS file entries

New Features
++++++++++++
- Integrate MSE notifications into PacketFence
## Delete branch after merge

NO

### CODE REVIEW

File: lib/pf/services/manager/httpd_mse.pm [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+extends 'pf::services::manager::httpd';
+
+has '+name' => (default => sub { 'httpd.mse' } );
+
+sub vhosts {
+    my ($self) = @_;
+    my @vhosts;
+    if ($management_network && defined($management_network->{'Tip'}) && $management_network->{'Tip'} ne '') {
+        push @vhosts, $management_network->{'Tip'};
```
Comment (fdurand) [inline lib/pf/services/manager/httpd_mse.pm]: Active/passive cluster will fail.

File: conf/httpd.conf.d/httpd.mse.tt.example [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        AuthType Basic
+        require valid-user
+    </Location>
+</VirtualHost>
+
+[% END %]
+
+Listen 127.0.0.1:[% port %]
+
+<VirtualHost 127.0.0.1:[% port %] >
```
Comment (fdurand) [inline conf/httpd.conf.d/httpd.mse.tt.example]: I don't think PacketFence will send to himself MSE notification

**Comment by fdurand** [comment]
It miss the iptable rule for the port 9292

File: conf/documentation.conf [hunk 3]
```diff
@@ -1405,6 +1425,13 @@ Set the max size of DHCP packets
 Do not change unless you know what you are doing
 EOT
 
+[advanced.extended_profile_filter_namespaces]
+type=list
```
Comment (fdurand) [inline conf/documentation.conf]: type=merged_list

File: conf/httpd.conf.d/httpd.mse.tt.example [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    ServerName [% server_name %]
+    DocumentRoot [% install_dir %]/html/pfappserver/lib
+    ErrorLog [% install_dir %]/logs/[% name %].error
+    CustomLog [% install_dir %]/logs/[% name %].access combined env=!is_collectd
+    SSLEngine on
+    Include [% var_dir %]/conf/ssl-certificates.conf
+    <Location / >
+        SetHandler modperl
+        PerlResponseHandler pf::WebAPI
+        PerlAuthenHandler pf::WebAPI::AuthenHandler
```
Comment (fdurand) [inline conf/httpd.conf.d/httpd.mse.tt.example]: Mse doesn't offer the way to give username and password

File: conf/httpd.conf.d/httpd.mse.tt.example [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+<VirtualHost [% vhost %]:[% port %] >
+    ServerName [% server_name %]
+    DocumentRoot [% install_dir %]/html/pfappserver/lib
+    ErrorLog [% install_dir %]/logs/[% name %].error
+    CustomLog [% install_dir %]/logs/[% name %].access combined env=!is_collectd
+    SSLEngine on
+    Include [% var_dir %]/conf/ssl-certificates.conf
+    <Location / >
+        SetHandler modperl
+        PerlResponseHandler pf::WebAPI
```
Comment (fdurand) [inline conf/httpd.conf.d/httpd.mse.tt.example]: pf::WebAPI::MSEHandler

File: lib/pf/node.pm [hunk 6]
```diff
@@ -23,6 +23,9 @@ use pf::log;
 use Readonly;
 use pf::StatsD::Timer;
 use pf::util::statsd qw(called);
+use pf::Redis;
```
Comment (fdurand) [inline lib/pf/node.pm]: useless

File: lib/pf/node.pm [hunk 7]
```diff
@@ -23,6 +23,9 @@ use pf::log;
...
+use pf::config qw(%Config);
```
Comment (fdurand) [inline lib/pf/node.pm]: useless

File: lib/pf/node.pm [hunk 8]
```diff
@@ -23,6 +23,9 @@ use pf::log;
...
+use JSON::MaybeXS;
```
Comment (fdurand) [inline lib/pf/node.pm]: useless

**Comment by fdurand** [comment]
It miss httpd.collector.tt in the packaging, still missing port 9292 in iptables

File: conf/httpd.conf.d/httpd.collector.tt.example [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+    Include [% var_dir %]/conf/ssl-certificates.conf
+    <Location /mse/ >
+        SetHandler modperl
+        PerlResponseHandler pf::WebAPI::MSEHandler
+    </Location>
+</VirtualHost>
+
+[% END %]
+
+Listen 127.0.0.1:[% port %]
```
Comment (fdurand) [inline conf/httpd.conf.d/httpd.collector.tt.example]: Do we really need to listen on 127.0.0.1 ?

File: conf/httpd.conf.d/httpd.collector.tt.example [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+<Directory />
+    AllowOverride None
+</Directory>
+
+<VirtualHost [% vhost %]:[% port %] >
+    ServerName [% server_name %]
+    DocumentRoot [% install_dir %]/html/pfappserver/lib
+    ErrorLog [% install_dir %]/logs/[% name %].error
+    CustomLog [% install_dir %]/logs/[% name %].access combined env=!is_collectd
+    SSLEngine on
```
Comment (fdurand) [inline conf/httpd.conf.d/httpd.collector.tt.example]: Remove ssl

File: conf/httpd.conf.d/httpd.collector.tt.example [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+<Directory />
+    AllowOverride None
+</Directory>
+
+<VirtualHost [% vhost %]:[% port %] >
+    ServerName [% server_name %]
+    DocumentRoot [% install_dir %]/html/pfappserver/lib
+    ErrorLog [% install_dir %]/logs/[% name %].error
+    CustomLog [% install_dir %]/logs/[% name %].access combined env=!is_collectd
```
Comment (fdurand) [inline conf/httpd.conf.d/httpd.collector.tt.example]: Remove it, it will produce useless io

File: conf/log.conf.d/httpd.collector.conf.example [hunk 12]
```diff
@@ -0,0 +1,19 @@
+### httpd.mse logger ###
```
Comment (fdurand) [inline conf/log.conf.d/httpd.collector.conf.example]: httpd.collector

File: conf/log.conf.d/httpd.collector.conf.example [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+log4perl.rootLogger = INFO, HTTPD_MSE
+
+### Categories ###
+# Below, you can specify different categories (based on package names) for different logging levels
+#log4perl.category.pf.SNMP = WARN
+
+### Logging for httpd.mse
+log4perl.appender.HTTPD_MSE                              = Log::Log4perl::Appender::File
+log4perl.appender.HTTPD_MSE.filename                     = /usr/local/pf/logs/httpd.mse.log
```
Comment (fdurand) [inline conf/log.conf.d/httpd.collector.conf.example]: httpd.collector.log

File: conf/log.conf.d/httpd.collector.conf.example [hunk 14]
```diff
@@ -0,0 +1,19 @@
+### httpd.mse logger ###
+log4perl.rootLogger = INFO, HTTPD_MSE
+
+### Categories ###
+# Below, you can specify different categories (based on package names) for different logging levels
+#log4perl.category.pf.SNMP = WARN
+
+### Logging for httpd.mse
+log4perl.appender.HTTPD_MSE                              = Log::Log4perl::Appender::File
```
Comment (fdurand) [inline conf/log.conf.d/httpd.collector.conf.example]: HTTPD_COLLECTOR

File: lib/pfconfig/namespaces/config/Pf.pm [hunk 15]
```diff
@@ -115,8 +115,10 @@ sub build_child {
             $Config{$group}{$item} = File::Spec->catfile( $log_dir, $Config{$group}{$item} );
         }
     }
-
-    $Config{trapping}{passthroughs} = [ split( /\s*,\s*/, $Config{trapping}{passthroughs} || '' ) ];
+    foreach my $val ("trapping.passthroughs") {
```
Comment (fdurand) [inline lib/pfconfig/namespaces/config/Pf.pm:118]: Not sure to understand why this is in the pull request ?

File: lib/pf/services/manager/httpd_collector.pm [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+extends 'pf::services::manager::httpd';
+
+has '+name' => (default => sub { 'httpd.collector' } );
+
+sub vhosts {
+    my ($self) = @_;
+    my @vhosts;
+    if ($management_network && defined($management_network->{'Tip'}) && $management_network->{'Tip'} ne '') {
+        push @vhosts, $management_network->{'Tip'};
```
Comment (fdurand) [inline lib/pf/services/manager/httpd_collector.pm]: https://github.com/inverse-inc/packetfence/commit/77a6089350486203c1116929466ce560813ffe9d

File: lib/pf/extended_node_data.pm [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+=head2 extended_node_get_data
+
+Get the node extended data
+
+=cut
+
+sub extended_node_get_data {
+    my ($mac, $namespace) = @_;
+    my $redis = pf::Redis->new(server => 'localhost:6379');
```
Comment (fdurand) [inline lib/pf/extended_node_data.pm]: 127.0.0.1

**Comment by fdurand** [comment]
Can you add a little example of a advance filter ?

File: lib/pfconfig/namespaces/config/Pf.pm (hunk 15 above)
Comment (rouzier) [inline lib/pfconfig/namespaces/config/Pf.pm:118]: This was a refactor done because of the extended namespace.
However we removed that in a later commit.

**Comment by jrouzierinverse** [comment]
Please re-review

File: conf/log.conf.d/httpd.collector.conf.example [hunk 18]
```diff
@@ -0,0 +1,19 @@
+### httpd.collector logger ###
+log4perl.rootLogger = INFO, HTTPD_COLLECTOR
+
+### Categories ###
+# Below, you can specify different categories (based on package names) for different logging levels
+#log4perl.category.pf.SNMP = WARN
+
+### Logging for httpd.mse
```
Comment (fdurand) [inline conf/log.conf.d/httpd.collector.conf.example]: httpd.mse => httpd.collector

File: addons/packages/packetfence.spec [hunk 19]
```diff
@@ -1070,6 +1070,8 @@ fi
                         /usr/local/pf/conf/httpd.conf.d/httpd.proxy.tt.example
 %config                 /usr/local/pf/conf/httpd.conf.d/httpd.webservices.tt
                         /usr/local/pf/conf/httpd.conf.d/httpd.webservices.tt.example
+%config                 /usr/local/pf/conf/httpd.conf.d/httpd.collector.tt
```
Comment (fdurand) [inline addons/packages/packetfence.spec:1073]: It miss the same thing for debian packaging (debian/packetfence.conffiles)

File: lib/pf/WebAPI/InitCollectorHandler.pm [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+pf::WebAPI::InitCollectorHandler
+
+=cut
+
+use strict;
+use warnings;
+
+use Apache2::RequestRec ();
+use pf::log;
```
Comment (fdurand) [inline lib/pf/WebAPI/InitCollectorHandler.pm:19]: useless

File: lib/pf/WebAPI/InitCollectorHandler.pm [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+use pf::Redis;
+use JSON::MaybeXS;
```
Comment (fdurand) [inline lib/pf/WebAPI/InitCollectorHandler.pm:21]: useless

File: lib/pf/web/httpd.collector_modperl_require.pl [hunk 22]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/perl
```
Comment (fdurand) [inline lib/pf/web/httpd.collector_modperl_require.pl:1]: File is define but not use anywhere (checkup.pm)

Comment (jrouzierinverse) [inline lib/pf/web/httpd.collector_modperl_require.pl:1]: It is used in 
https://github.com/jrouzierinverse/packetfence/blob/feature/mse-integration/conf/httpd.conf.d/httpd.collector.tt.example#L108

**Comment by louismunro** [comment]
review status please?

**Pull request merged**